README.md
## musiccharts

Generate music chart PDFs in multiple keys plus Nashville Number System (NNS).  User provides a plain text document in NNS adhering to the formatting standards, output is formatted PDF documents in all keys (default) or only user-requested keys.

### Environment Requirements
* Tested on python3.7, PyLaTeX 1.4.1

### Setup
* Download repo and run `pip3 install .` from the folder.

### CLI Usage & Examples
```
> musiccharts test_lyrics2.txt -k A,Bb,B
> musiccharts test_lyrics2.txt --tex --name "newfile"

> musiccharts
usage: musiccharts [-h] [-k] [-n] [-s] [--tex] [input [input ...]]

Description...

positional arguments:
  input         Input file

optional arguments:
  -h, --help   show this help message and exit
  -k, --key    Desired keys, default is 
  -n, --name   Destination filename
  -s, --size   Available font sizes: [8, 8.5, 9, 9.5, 10, 10.5, 11, 11.5, 12, 12.5, 13,
               13.5, 14, 14.5, 15, 15.5, 16, 16.5, 17, 17.5, 18, 18.5, 19, 19.5, 20]
  --tex        Output LaTeX file in addition to PDFs
```

### Design
* Title formatting tag: `Title{Lorem Ipsum}`
* Right-adjusted intro labels: **`Intro:`**, **`Intro/Interlude:`**, **`Intro/Outro:`**
* Section labels: **`Bridge:`**, **`Chorus:`**, **`Interlude:`**, **`Outro:`**, **`Prechorus:`**, **`Tag:`**, **`Verse:`**
* All chords must have at least 2 spaces between them
* Margins (0.25in), superscript scaling (75%), and other parameters are adjustable in `data_definitions.py`

### Chord Symbols
```
Chord regex - (^|\n| )([#b]?)(\d)([0-9susb#+-/oø△Δ]*)( |$|\n)
sus = suspended
-   = minor
-7  = minor 7th
△   = major 7th
ø   = half diminished 7th
o   = full-diminished 7th
+   = augmented
```

### Feedback
* This tool is in alpha with fairly limited testing so far
* Feedback, bug fixes, or improvements are welcomed
* Please submit an Issue via GitHub, drop an email to github@jamesw.us


CHANGELOG.md
# Change Log

## 0.1.0

### Alpha release

* Render charts in all keys + NNS
* Whitespace adjustment for chord character counts (e.g going from 5sus -> C#sus)
* Whitespace padding for scaled superscripts to improve character alignment
* CLI flags for all keys, options, filenames, font sizes, and tex file debug mode
* Basic error checking for keys, and invalid chord inversions
* Sample input file containing a wide set of chord syntax examples 

### Planned for future releases

* More robust syntax checks
* Checks for text exceeding page width/length

musiccharts/cli.py
#!/usr/bin/env python3

"""
Take in a music chart in raw text form.
Output 13 PDF's:
- One in the Nashville Number System
- One in each of the lettered keys

Dependencies:
- pylatex package
"""

import argparse
import os
import re
import signal
import sys

from musiccharts import __version__
from musiccharts import data_definitions as vars
from musiccharts.data_processing import (format_line, process_document,
                                         validate_keys)
from musiccharts.read_file import read_file_contents


def main():
    """Execute main function."""

    # ArgParse setup
    formatter_class = argparse.RawDescriptionHelpFormatter
    description = "Description..."
    epilog = f"Version: {__version__}\n "
    parser = argparse.ArgumentParser(
        formatter_class=formatter_class, description=description, epilog=epilog
    )
    parser.add_argument("input", nargs="*", help="Input file")
    parser.add_argument(
        "-k",
        "--key",
        default="NNS,Ab,A,Bb,B,C,Db,D,Eb,E,F,Gb,G",
        dest="keys",
        help='Desired keys, default is "NNS,Ab,A,Bb,B,C,Db,D,Eb,E,F,Gb,G"',
        metavar="",
        type=str,
    )
    parser.add_argument(
        "-n",
        "--name",
        dest="dest_filename",
        help="Destination filename",
        metavar="",
        type=str,
    )
    parser.add_argument(
        "-s",
        "--size",
        default=vars.BASE_FONT_SIZE,
        choices=vars.ALL_FONT_SIZES,
        dest="font_size",
        help=f"Available font sizes: {vars.ALL_FONT_SIZES}",
        metavar="",
        type=float,
    )
    parser.add_argument(
        "--tex",
        dest="keep_tex",
        action="store_false",
        help="Output LaTeX file in addition to PDFs",
    )
    args = parser.parse_args()
    if len(args.input) > 1:
        print("\nError: more than one source file provided for processing!\n")
        sys.exit(1)
    elif len(args.input) < 1:
        parser.print_help()
        sys.exit(1)
    else:
        src_filename = args.input[0]

    # Process provided filename
    file_contents = read_file_contents(filename=src_filename)

    if args.dest_filename:
        dest_filename = args.dest_filename
    else:
        dest_filename = re.sub(r"(.*/)*([^.]*)(\..*)", r"\2", src_filename)

    # Save and validate user-provided keys
    keys = args.keys.split(",")
    validate_keys(keys)

    # Process formatting for each individual line of text
    for key in keys:
        formatted_lines = []
        chord_errors = []
        i = 0
        for line in file_contents:
            results = format_line(line=line, line_num=i, key=key, debug=False)
            formatted_lines.append(results.get("edits"))
            chord_errors.extend(results.get("errors"))
            i += 1

        # Display specific invalid chord errors
        if len(chord_errors) > 0:
            print("\nError: invalid chord syntax found")
            print("----------------------------------")
            for match in chord_errors:
                print(
                    f"Line {str(match.get('line_num')).ljust(3)}: {match.get('chord')}"
                )
            print("")
            sys.exit(1)

        # Create the document
        process_document(
            dest_filename=f"{dest_filename} ({key})",
            path=os.getcwd(),
            file_contents=formatted_lines,
            font_size=args.font_size,
            keep_tex=args.keep_tex,
        )


# Call main function
if __name__ == "__main__":
    main()


musiccharts/read_file.py
#!/usr/bin/env python3

"""
Handles reading in data from a plain-text file.
"""

import os
import sys

import musiccharts.data_definitions as vars


def read_file_contents(filename):
    """
    Read file line by line and return contents.

    Args:
        filename: (str) input file path, either relative or absolute

    Returns:
        list of str: lines from file

    """
    file_contents = []

    # Determine whether to treat filename as relative or absolute
    if not os.path.isfile(filename):
        filename = os.path.join(os.getcwd(), filename)

    # Open file object as read-only
    try:
        file_object = open(filename, "r")
    except FileNotFoundError:
        print(f"File not found: {filename}\nExiting...")
        sys.exit(0)

    # Read file contents line by line
    for line in file_object:
        file_contents.append(line.rstrip())

    # Close out the file
    try:
        file_object.close()
    except Exception:
        pass

    # Return a list of strings
    return file_contents


musiccharts/data_processing.py
#!/usr/bin/env python3

"""
Handles chord regex substitution and LaTeX processing/output
"""

import os
import re
import sys

import musiccharts.data_definitions as vars
from pylatex import Command, Document, Section, Subsection, UnsafeCommand
from pylatex.basic import NewLine
from pylatex.utils import NoEscape, italic, verbatim


def validate_keys(keys):
    """
    Input a list comma-separated list of keys, fail if invalid keys present

    Args:
        keys: (list of string) e.g. ['A','B','C','NNS']
    """
    bad_keys = []
    for key in keys:
        if key not in list(vars.CHORDS.keys()):
            bad_keys.append(key)
    if len(bad_keys) > 0:
        print(f"\nError - invalid keys found: {','.join(bad_keys)}\n")
        sys.exit(1)


def format_line(line, line_num, key, debug=False):
    """
    Input a line of text and process any of the following:
      - NNS chord notation. symbols get superscripted, plus whitespace adjustment
      - Title, which must be wrapped in "Title{<title text>}"
      - Section labels (Intro, Verse, Chorus, etc)

    Args:
        line: single line of text containing mixed lyrics, labels, chords
        line_num: used only for outputting chord syntax errors
        key: which key to transpose into
        debug: enable debug mode to print text lines pre/post changes

    Returns:
        string: formatted line of text
    """
    if debug:
        print(f"PRE: {line}")

    # Format intro in the title line
    intro_matches = vars.INTRO_REGEX.findall(line)
    for item in intro_matches:
        line = line.replace(item, r"\hfill{\textbf{" + item + "}}")

    # Format title
    title_matches = vars.TITLE_REGEX.findall(line)
    if len(title_matches) > 0:
        line = r"{\bigtitle{" + title_matches[0][1] + "}}" + title_matches[0][3]

    # Add bold for verse/chorus labels
    label_matches = vars.LABEL_REGEX.findall(line)
    for item in label_matches:
        line = line.replace(item, r"\textbf{" + item + "}")

    # Regex searches for chord regex
    chord_matches = vars.CHORD_REGEX1.findall(line.replace("(", "").replace(")", ""))
    chords = []
    chord_errors = []
    for item in chord_matches:
        original_chord = ""
        edited_chord = ""
        item = list(item)
        remove_full_space = 0
        remove_partial_space = 0
        for i in range(0, 5):
            original_chord += item[i]
            # For NNS, pre-chord name/number accidentals (#/b) get superscripted, left-padded
            # For regular keys, convert the provided NNS [b|#][0-7] to [A-G][b|#], no superscript
            if i == 1:
                if key == "NNS":
                    if item[i] not in ["", " ", "\n"]:
                        spacing = len(item[i].replace("△", "")) * " "
                        edited_chord += r"\ts{{\thin" + spacing + "}" + item[i] + "}"
                else:
                    translation = vars.CHORDS[key][f"{item[1]}{item[2]}"]
                    if len(translation) > 1:
                        remove_full_space += 1
                    edited_chord += translation
            elif i == 2:
                if key == "NNS":
                    translation = vars.CHORDS[key][item[i]]
                    if len(translation) > 1:
                        remove_full_space += 1
                    edited_chord += translation
            # Post-chord name/number symbols/digits get superscripted, right-padded
            elif i == 3:
                if item[i] not in ["", " ", "\n"]:
                    # Check for invalid chord inversions
                    # Example: 1/3, 1sus/3 are a valid chord3, 1/34 1/sus3 are not
                    if "/" in item[i]:
                        splits = item[i].split("/")
                        if (
                            len(vars.CHORD_REGEX2.findall(f"/{splits[1]}")) == 0
                            or len(vars.CHORD_REGEX3.findall(f"/{splits[1]}")) == 1
                            or item[i].count("/") > 1
                        ):
                            chord_errors.append(
                                {"chord": "".join(item), "line_num": line_num}
                            )
                            continue
                        translation = vars.CHORDS[key][splits[1]]
                        if len(translation) > len(splits[1]):
                            remove_full_space += 1
                        item[i] = f"{splits[0]}/{translation}"
                    count = len(item[i].replace("△", "")) - remove_partial_space
                    spacing = count * " "
                    edited_chord += r"\ts{" + item[i] + r"{\thin" + spacing + "}}"
            else:
                edited_chord += item[i]
        if remove_full_space == 1:
            original_chord = f"{original_chord} "
        if (original_chord, edited_chord) not in chords:
            chords.append((original_chord, edited_chord))

    # Replace raw chords with LaTeX-formatted chords
    for chord in chords:
        line = f"{line} ".replace(chord[0], chord[1]).rstrip()

    if debug:
        print(f"POST: {line}")

    return {"edits": NoEscape(line), "errors": chord_errors}


def process_document(dest_filename, path, file_contents, font_size, keep_tex):
    """
    Create TEX document based on file input and given parameters.

    Args:
        dest_filename: name of the file without extension
        path: destination folder path
        file_contents: plain text input from file
        base_font_size: user-provided font size or default
        keep_tex: generate TEX file in addition to PDF
    """
    print(f"FILENAME:  {dest_filename}.pdf")
    print(f"FONT SIZE: {font_size}")
    # Document type
    doc = Document(
        default_filepath=dest_filename,
        documentclass="extarticle",
        document_options=[f"{vars.BASE_FONT_SIZE}pt"],
        fontenc=None,
        indent=False,
        inputenc="utf8",
        lmodern=False,
        page_numbers=False,
        textcomp=False,
    )
    font_scale_ratio = round(font_size / vars.BASE_FONT_SIZE, 2)

    # PyLatex automatically adds unneeded additional line break chars
    # Content separator has to be redefined to avoid breaking some commands
    doc.content_separator = "\n"

    # 'fontspec' package required for font scaling
    doc.preamble.append(Command(command="usepackage", arguments="fontspec,xfp"))
    # Paragraph indent settings
    doc.preamble.append(Command(command="usepackage", arguments="parskip"))
    # Font selection
    doc.preamble.append(Command(command="usepackage", arguments="courier"))
    # 'fancyvrb' allows use of monospace font with superscripts
    doc.preamble.append(Command(command="usepackage", arguments="fancyvrb"))
    # Font selection
    doc.preamble.append(Command(command="usepackage", arguments="fvextra"))
    # Use geometry to set margins
    doc.preamble.append(
        Command(
            command="usepackage",
            arguments="geometry",
            options=f"margin={vars.ALL_MARGINS}in",
        )
    )
    # 'newunicodechar' package required to map specific chars: △, ø
    doc.preamble.append(
        Command(
            command="usepackage",
            arguments="newunicodechar",
            options="verbose",
        )
    )
    # Set the monospace font
    doc.preamble.append(NoEscape(r"\setmonofont{Courier New}"))

    # Setup for font scaling
    doc.preamble.append(NoEscape(r"\makeatletter"))
    doc.preamble.append(NoEscape(r"\newcommand{\scalefont}[1]{"))
    doc.preamble.append(NoEscape(r"    \edef\scale@fontsize{\fpeval{#1*\f@size}}"))
    doc.preamble.append(
        NoEscape(r"    \edef\scale@fontbaselineskip{\fpeval{1.2*\scale@fontsize}}")
    )
    doc.preamble.append(
        NoEscape(r"    \fontsize{\scale@fontsize}{\scale@fontbaselineskip}\selectfont}")
    )
    doc.preamble.append(NoEscape(r"\makeatother"))

    # Redefine font size for superscript
    doc.preamble.append(
        NoEscape(
            r"\renewcommand{\textsuperscript}[1]{\raisebox{"
            + str(vars.SUPERSCRIPT_RAISE)
            + r"ex}{\scalefont{"
            + str(vars.SUPERSCRIPT_SCALER)
            + r"}#1}}"
        )
    )

    # Set superscripts to use monospace font as default font
    doc.preamble.append(NoEscape(r"\renewcommand{\familydefault}{\ttdefault}"))

    # Set short custom command for superscript
    doc.preamble.append(NoEscape(r"\newcommand{\ts}{\textsuperscript}"))

    # Choose font scaling for odd vs even font size
    doc.preamble.append(
        NoEscape(
            r"\newfontfamily{\thin}[Scale="
            + str(vars.SUPERSCRIPT_FILLER_SCALER)
            + r"]{Courier New}"
        )
    )

    # Define font/size for title
    title_font_cmd = [
        r"\newfontfamily\bigtitle[SizeFeatures={Size=",
        str(vars.TITLE_SIZE) + "}]{",
        str(vars.TITLE_FONT) + "}",
    ]
    doc.preamble.append(NoEscape("".join(title_font_cmd)))

    # Define extra unicode chars
    doc.preamble.append(NoEscape(r"\newunicodechar{△}{$\bigtriangleup$}"))
    doc.preamble.append(NoEscape(r"\newunicodechar{ø}{$\o$}"))

    # Use 'Verbatim' for document body to allow use of '\textsuperscript{}'
    doc.append(
        NoEscape(
            r"\begin{Verbatim}[fontsize=\scalefont{"
            + str(font_scale_ratio)
            + r"},commandchars=\\\{\}]"
        )
    )
    for line in file_contents:
        doc.append(line)
    doc.append(NoEscape(r"\end{Verbatim}"))
    doc.generate_pdf(clean_tex=keep_tex, compiler_args=["--xelatex"])
    os.remove(f"{os.getcwd()}/{dest_filename}.xdv")


musiccharts/data_definitions.py
#!/usr/bin/env python3
import re

ALL_MARGINS = 0.25  # In inches
ALL_FONT_SIZES = [
    8,
    8.5,
    9,
    9.5,
    10,
    10.5,
    11,
    11.5,
    12,
    12.5,
    13,
    13.5,
    14,
    14.5,
    15,
    15.5,
    16,
    16.5,
    17,
    17.5,
    18,
    18.5,
    19,
    19.5,
    20,
]
BASE_FONT_SIZE = 14  # In points
SUPERSCRIPT_RAISE = 0.75
SUPERSCRIPT_SCALER = 0.75
SUPERSCRIPT_FILLER_SCALER = (1 - SUPERSCRIPT_SCALER) / SUPERSCRIPT_SCALER
TITLE_SIZE = 20  # In points
TITLE_FONT = "Futura"

# Regex compilation for chart formatting
CHORD_REGEX1 = re.compile(r"(^|\n| )([#b]?)(\d)([0-9susb#+-oø△/]*)( |$|\n|\))")
CHORD_REGEX2 = re.compile(r"/([#b]?)([0-9])( |$|\n|\))")
CHORD_REGEX3 = re.compile(r"/([#b]?)([8-9])( |$|\n|\))")
INTRO_REGEX = re.compile(r"(?:Intro|Intro/Interlude|Intro/Outro):")
LABEL_REGEX = re.compile(
    r"(?:Bridge|Chorus|Interlude|Outro|Prechorus|Tag|Verse|Ending)\s?[0-9]?:"
)
TITLE_REGEX = re.compile(r"(^Title{)([^}]*)(})(.*)")

# Predefine chords per-key for transposition
CHORDS = {}
CHORDS["NNS"] = {
    "1": "1",
    "#1": "#1",
    "b2": "b2",
    "2": "2",
    "#2": "#2",
    "b3": "b3",
    "3": "3",
    "4": "4",
    "#4": "#4",
    "b5": "b5",
    "5": "5",
    "#5": "#5",
    "b6": "b6",
    "6": "6",
    "#6": "#6",
    "b7": "b7",
    "7": "7",
}
CHORDS["A"] = {
    "1": "A",
    "#1": "A#",
    "b2": "Bb",
    "2": "B",
    "#2": "C",
    "b3": "C",
    "3": "C#",
    "4": "D",
    "#4": "D#",
    "b5": "Eb",
    "5": "E",
    "#5": "F",
    "b6": "F",
    "6": "F#",
    "#6": "G",
    "b7": "G",
    "7": "G#",
}
CHORDS["A#"] = {
    "1": "A#",
    "#1": "B",
    "b2": "B",
    "2": "C",
    "#2": "C#",
    "b3": "Db",
    "3": "D",
    "4": "D#",
    "#4": "E",
    "b5": "E",
    "5": "F",
    "#5": "F#",
    "b6": "Gb",
    "6": "G",
    "#6": "G#",
    "b7": "Ab",
    "7": "A",
}
CHORDS["Bb"] = {
    "1": "Bb",
    "#1": "B",
    "b2": "B",
    "2": "C",
    "#2": "C#",
    "b3": "Db",
    "3": "D",
    "4": "D#",
    "#4": "E",
    "b5": "E",
    "5": "F",
    "#5": "F#",
    "b6": "Gb",
    "6": "G",
    "#6": "G#",
    "b7": "Ab",
    "7": "A",
}
CHORDS["B"] = {
    "1": "B",
    "#1": "C",
    "b2": "C",
    "2": "C#",
    "#2": "D",
    "b3": "D",
    "3": "D#",
    "4": "E",
    "#4": "F",
    "b5": "F",
    "5": "F#",
    "#5": "G",
    "b6": "G",
    "6": "G#",
    "#6": "A",
    "b7": "A",
    "7": "A#",
}
CHORDS["C"] = {
    "1": "C",
    "#1": "C#",
    "b2": "Db",
    "2": "D",
    "#2": "D#",
    "b3": "Eb",
    "3": "E",
    "4": "F",
    "#4": "F#",
    "b5": "Gb",
    "5": "G",
    "#5": "G#",
    "b6": "Ab",
    "6": "A",
    "#6": "A#",
    "b7": "Bb",
    "7": "B",
}
CHORDS["C#"] = {
    "1": "C#",
    "#1": "D",
    "b2": "D",
    "2": "D#",
    "#2": "E",
    "b3": "E",
    "3": "F",
    "4": "F#",
    "#4": "G",
    "b5": "G",
    "5": "G#",
    "#5": "A",
    "b6": "A",
    "6": "A#",
    "#6": "B",
    "b7": "B",
    "7": "C",
}
CHORDS["Db"] = {
    "1": "Db",
    "#1": "D",
    "b2": "D",
    "2": "Eb",
    "#2": "E",
    "b3": "E",
    "3": "F",
    "4": "Gb",
    "#4": "G",
    "b5": "G",
    "5": "Ab",
    "#5": "A",
    "b6": "A",
    "6": "Bb",
    "#6": "B",
    "b7": "B",
    "7": "C",
}
CHORDS["D"] = {
    "1": "D",
    "#1": "D#",
    "b2": "Eb",
    "2": "E",
    "#2": "F",
    "b3": "F",
    "3": "F#",
    "4": "G",
    "#4": "G#",
    "b5": "Ab",
    "5": "A",
    "#5": "A#",
    "b6": "Bb",
    "6": "B",
    "#6": "C",
    "b7": "C",
    "7": "C#",
}
CHORDS["D#"] = {
    "1": "D#",
    "#1": "E",
    "b2": "E",
    "2": "F",
    "#2": "F#",
    "b3": "Gb",
    "3": "G",
    "4": "Ab",
    "#4": "A",
    "b5": "A",
    "5": "A#",
    "#5": "B",
    "b6": "B",
    "6": "C",
    "#6": "C#",
    "b7": "Db",
    "7": "D",
}
CHORDS["Eb"] = {
    "1": "Eb",
    "#1": "E",
    "b2": "E",
    "2": "F",
    "#2": "F#",
    "b3": "Gb",
    "3": "G",
    "4": "Ab",
    "#4": "A",
    "b5": "A",
    "5": "Bb",
    "#5": "B",
    "b6": "B",
    "6": "C",
    "#6": "C#",
    "b7": "Db",
    "7": "D",
}
CHORDS["E"] = {
    "1": "E",
    "#1": "F",
    "b2": "F",
    "2": "F#",
    "#2": "G",
    "b3": "G",
    "3": "G#",
    "4": "A",
    "#4": "A#",
    "b5": "Bb",
    "5": "B",
    "#5": "C",
    "b6": "C",
    "6": "C#",
    "#6": "D",
    "b7": "D",
    "7": "D#",
}
CHORDS["F"] = {
    "1": "F",
    "#1": "F#",
    "b2": "Gb",
    "2": "G",
    "#2": "G#",
    "b3": "Ab",
    "3": "A",
    "4": "Bb",
    "#4": "B",
    "b5": "B",
    "5": "C",
    "#5": "C#",
    "b6": "Db",
    "6": "D",
    "#6": "D#",
    "b7": "Eb",
    "7": "E",
}
CHORDS["F#"] = {
    "1": "F#",
    "#1": "G",
    "b2": "G",
    "2": "G#",
    "#2": "A",
    "b3": "A",
    "3": "A#",
    "4": "B",
    "#4": "C",
    "b5": "C",
    "5": "C#",
    "#5": "D",
    "b6": "D",
    "6": "D#",
    "#6": "E",
    "b7": "E",
    "7": "F",
}
CHORDS["Gb"] = {
    "1": "Gb",
    "#1": "G",
    "b2": "G",
    "2": "Ab",
    "#2": "A",
    "b3": "A",
    "3": "Bb",
    "4": "B",
    "#4": "C",
    "b5": "C",
    "5": "Db",
    "#5": "D",
    "b6": "D",
    "6": "Eb",
    "#6": "E",
    "b7": "E",
    "7": "F",
}
CHORDS["G"] = {
    "1": "G",
    "#1": "G#",
    "b2": "Ab",
    "2": "A",
    "#2": "A#",
    "b3": "Bb",
    "3": "B",
    "4": "C",
    "#4": "C#",
    "b5": "Db",
    "5": "D",
    "#5": "D#",
    "b6": "Eb",
    "6": "E",
    "#6": "F",
    "b7": "F",
    "7": "F#",
}
CHORDS["G#"] = {
    "1": "G#",
    "#1": "A",
    "b2": "A",
    "2": "A#",
    "#2": "B",
    "b3": "B",
    "3": "C",
    "4": "C#",
    "#4": "D",
    "b5": "D",
    "5": "D#",
    "#5": "E",
    "b6": "E",
    "6": "F",
    "#6": "F#",
    "b7": "Gb",
    "7": "G",
}
CHORDS["Ab"] = {
    "1": "Ab",
    "#1": "A",
    "b2": "A",
    "2": "Bb",
    "#2": "B",
    "b3": "B",
    "3": "C",
    "4": "Db",
    "#4": "D",
    "b5": "D",
    "5": "Eb",
    "#5": "E",
    "b6": "E",
    "6": "F",
    "#6": "F#",
    "b7": "Gb",
    "7": "G",
}
